### PR TITLE
Add Psalm, with a lot of issues suppressed

### PR DESCRIPTION
**Pull request by muglug**
This PR adds the necessary fixes to get Psalm (mostly) working.

The idea here is that you'd remove issues (e.g. `InvalidReturnType`) when you want to make some headway.

***Edit***: errors are all accounted for

### CODE REVIEW

File: src/Database/Query.php [hunk 1]
```diff
@@ -179,7 +179,7 @@ public function getConnection()
      *
      * @deprecated 3.4.0 Use setConnection()/getConnection() instead.
      * @param \Cake\Database\Connection|null $connection Connection instance
-     * @return $this|\Cake\Database\Connection
+     * @return static|\Cake\Database\Connection
```
Comment (dereuromark) [inline src/Database/Query.php]: Wasn't there a reason (IDEs) that we stuck to the $this for fluent interface and non static?

File: src/ORM/Behavior/TimestampBehavior.php [hunk 2]
```diff
@@ -133,7 +133,7 @@ public function implementedEvents()
      *
      * @param \DateTime|null $ts Timestamp
      * @param bool $refreshTimestamp If true timestamp is refreshed.
-     * @return \DateTime
+     * @return Time
```
Comment (dereuromark) [inline src/ORM/Behavior/TimestampBehavior.php]: FQCN

File: src/Database/Query.php (hunk 1 above)
Comment (muglug) [reply src/Database/Query.php]: I can fix the code in Psalm that demands this change (only `@return $this` is currently permitted). No good reason, though I do prefer `static` a little.

Comment (dereuromark) [reply src/Database/Query.php]: Afaik static broke the fluent interface approach in IDEs. We would have to test if this change is good.
The deprecated methods should not be in use anymore though, so it most likely won't be too big of a deal either way. And for future methods we sure always try to only use true fluent typehinting (no getter mixin).

Comment (dereuromark) [reply src/Database/Query.php]: But I do see a few methods that are not even deprecated yet, they should be separated into their getter/setter part first IMO.

**Comment by muglug** [comment]
Also should I have Psalm run on the 5.6 build too (because it can)

**Comment by dereuromark** [comment]
> Also should I have Psalm run on the 5.6 build too (because it can)

I don't think that is worth it. We are now all more looking into 7.1+ direction rather than back.

File: src/Database/Query.php (hunk 1 above)
Comment (muglug) [reply src/Database/Query.php]: it would surprise me if IDEs (i.e. PHPStorm) didn't support `@return static` at this point.

Comment (lorenzo) [reply src/Database/Query.php]: I feel like this changes back and forth every other month

File: src/Utility/Text.php [hunk 3]
```diff
@@ -553,6 +553,7 @@ public static function stripLinks($text)
      * @param string $text String to truncate.
      * @param int $length Length of returned string, including ellipsis.
      * @param array $options An array of options.
+     * @psalm-param array{ellipsis?: string, exact?: bool} $options
```
Comment (lorenzo) [inline src/Utility/Text.php]: I like this

File: psalm.xml [hunk 4]
```diff
@@ -0,0 +1,185 @@
+<?xml version="1.0"?>
+<psalm
+    totallyTyped="false"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="/Users/matthewbrown/Desktop/vimeo/git/cakephp/vendor/vimeo/psalm/src/config.xsd"
```
Comment (markstory) [inline psalm.xml]: This location won't exist on most people's computers.

Comment (muglug) [reply psalm.xml]: works on mine. Production-ready code.

File: src/Controller/Component/AuthComponent.php [hunk 5]
```diff
@@ -880,7 +880,7 @@ public function constructAuthenticate()
      *
      * @param \Cake\Auth\Storage\StorageInterface|null $storage Sets provided
      *   object as storage or if null returns configured storage object.
-     * @return \Cake\Auth\Storage\StorageInterface|\Cake\Core\InstanceConfigTrait|null
```
Comment (ADmad) [inline src/Controller/Component/AuthComponent.php]: Removing `\Cake\Core\InstanceConfigTrait` makes `phpstan` unhappy.
https://travis-ci.org/cakephp/cakephp/jobs/356494536#L732

File: composer.json [hunk 6]
```diff
@@ -42,7 +42,8 @@
     },
     "require-dev": {
         "phpunit/phpunit": "^5.7.14|^6.0",
-        "cakephp/cakephp-codesniffer": "^3.0"
+        "cakephp/cakephp-codesniffer": "^3.0",
+        "vimeo/psalm": "^1.0"
```
Comment (ADmad) [inline composer.json]: Since `psalm` is being added as dependency in `travis.yml` itself, adding is here is not necessary. Most contributors aren't going to run it locally anyway.

File: src/Controller/Component/AuthComponent.php (hunk 5 above)
Comment (muglug) [reply src/Controller/Component/AuthComponent.php]: This is *definitely* an off-label use of traits, treating them like interfaces. PHP won't allow traits as return types, and I don't see why they should be allowed in a union type.

Reminds me of [this RFC](https://wiki.php.net/rfc/traits-with-interfaces), except without the interface.

I can add `@method` annotations to `StorageInterface` to make PHPStan happy.

File: composer.json (hunk 6 above)
Comment (muglug) [reply composer.json]: Absolutely. Will remove

**Comment by ADmad** [comment]
> ERROR: UndefinedMethod - src/Auth/ControllerAuthorize.php:93:22 - Method Cake\Controller\Controller::isAuthorized does not exist
        return (bool)$this->_Controller->isAuthorized($user);

This is a weird one as the `Controller` doesn't have an `isAuthorized()` method but developer is expected to add on in a child class extending `Controller` when using this `ControllerAuthorize` authorization adapter.

> ERROR: UndefinedMethod - src/Collection/Iterator/FilterIterator.php:71:21 - Method Traversable::getInnerIterator does not exist
        $iterator = $filter->getInnerIterator();

Not too sure about this. Perhaps @lorenzo can help.

> ERROR: AssignmentToVoid - src/Controller/Component/AuthComponent.php:386:9 - Cannot assign $result to type void
        $result = $auth->unauthenticated($controller->request, $response);

The `unauthenticated()` method of authentication adapters can either return `Cake\Http\Response` instance or return void. So maybe updating return types for `BaseAuthenticate::unauthenticated()` can help.

ERROR: InvalidPropertyAssignmentValue - src/Controller/Component/CookieComponent.php:127:30 - $this->request with declared type 'Cake\Http\ServerRequest' cannot be assigned type 'Zend\Diactoros\ServerRequest'
            $this->request = ServerRequestFactory::fromGlobals();

`CookieComponent` is deprecated so I don't think it's worth tackling this.

> ERROR: UndefinedMethod - src/Core/InstanceConfigTrait.php:187:16 - Method Cake\View\StringTemplate::getconfig does not exist
        return $this->getConfig($key);

This one seems like a bug in `psalm`. `StringTemplate` aliases `InstanceConfigTrait::getConfig()` as `get()`.
https://github.com/cakephp/cakephp/blob/3.next/src/View/StringTemplate.php#L33

> ERROR: NullReference - src/Database/Driver.php:334:16 - Cannot call method lastInsertId on null variable $this->_connection
        return $this->_connection->lastInsertId($table, $column);

The `$this->connect();` at start of method is supposed to assign a connection instance to `$this->_connection`.

> ERROR: InvalidArrayOffset - src/I18n/Parser/MoFileParser.php:130:17 - Cannot access value on variable $messages[$singularId]['_context'] using offset value of '_context', expecting int
                $messages[$singularId]['_context'][$context] = $singular;

> ERROR: InvalidArrayOffset - src/I18n/Parser/MoFileParser.php:132:21 - Cannot access value on variable $messages[$pluralId]['_context'] using offset value of '_context', expecting int
                    $messages[$pluralId]['_context'][$context] = $plurals;

No idea about these.

File: psalm.xml [hunk 7]
```diff
@@ -0,0 +1,185 @@
+<?xml version="1.0"?>
+<psalm
+    totallyTyped="false"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="vendor/vimeo/psalm/src/config.xsd"
```
Comment (ADmad) [inline psalm.xml]: Will this work if one is using `psalm.phar` instead of having it as a composer dependency?

**Comment by muglug** [comment]
> The $this->connect(); at start of method is supposed to assign a connection instance to $this->_connection.

Yeah but there's a check for `PDO` and a return. So you have some sort of non-PDO connection?

File: psalm.xml (hunk 7 above)
Comment (muglug) [reply psalm.xml]: the schemaLocation is just so IDEs can do autocomplete on the XML.

File: src/Controller/Component/AuthComponent.php (hunk 5 above)
Comment (ADmad) [reply src/Controller/Component/AuthComponent.php]: I am happy with whatever solution which makes both `phpstan` and `psalm` happy.

Mind you travis runs phpstan at level 2 but we are slowing working towards having it pass level 3. I hope the changes you make for `psalm` don't increases the errors reported for phpstan level 3. :slightly_smiling_face:

File: composer.json [hunk 8]
```diff
@@ -65,7 +65,6 @@
             "TestPluginTwo\\": "tests/test_app/Plugin/TestPluginTwo/src/",
             "Company\\TestPluginThree\\": "tests/test_app/Plugin/Company/TestPluginThree/src/",
             "Company\\TestPluginThree\\Test\\": "tests/test_app/Plugin/Company/TestPluginThree/tests/",
-            "ParentPlugin\\": "tests/test_app/Plugin/ParentPlugin/src/",
```
Comment (ADmad) [inline composer.json]: Why this change?

File: psalm.xml (hunk 7 above)
Comment (ADmad) [reply psalm.xml]: Okay.

File: composer.json (hunk 8 above)
Comment (muglug) [reply composer.json]: checkout out `master` branch composer.json by mistake. Fixed

**Comment by ADmad** [comment]
> Yeah but there's a check for PDO and a return. So you have some sort of non-PDO connection?

Nope, in case of `Driver` and child classes it will always be `PDO` instance.

File: src/Auth/BaseAuthenticate.php [hunk 9]
```diff
@@ -239,7 +239,7 @@ public function getUser(ServerRequest $request)
      *
      * @param \Cake\Http\ServerRequest $request A request object.
      * @param \Cake\Http\Response $response A response object.
-     * @return void
+     * @return \Cake\Http\Response|null
```
[the remark this answers is not in the document]
Comment (muglug) [reply src/Auth/BaseAuthenticate.php]: This was fixed (`return null` added)

File: src/Controller/Component/AuthComponent.php (hunk 5 above)
Comment (muglug) [reply src/Controller/Component/AuthComponent.php]: > I hope the changes you make for psalm don't increases the errors reported for phpstan level 3

I can't imagine how they would.

**Comment by ADmad** [comment]
This looks good now expect for the `$this` to `static` change.

**Comment by muglug** [comment]
@ADmad I've reverted the `static` changes. They keyword still sits in `DateFormatTrait`, but it wasn't me what did that.

**Review by ADmad — APPROVED** [review]
